File: Merge-Sort/js/leetcode_lc_51.js
const arr = []

function initData(count){
    for(let i = 0 ; i < count ; i++){
        arr.push(Math.floor(Math.random()*50-20))
    }
}

initData(5)

console.log(arr)

/**
 * @param {number[]} nums
 * @return {number}
 */
var reversePairs = function(nums) {
    let count = 0
    mergeSort(nums)

    function mergeSort(p){
        if(p.length < 2) return p
        const mid = Math.floor(p.length/2),
            left = p.slice(0,mid),
            right = p.slice(mid)
        return merge(mergeSort(left),mergeSort(right))
    }

    function merge(left,right){
        const res = []
        let i = j = 0
        while(i < left.length && j < right.length){  
            if(left[i] <= right[j]){
                res.push(left[i])
                i++
            }else{
                for(let idx = i; idx<left.length; idx++){
                    console.log("逆序对",left[i],right[j])
                }
                res.push(right[j])
                j++
                // 归并升序排序，因为左侧的数组和右侧数组都是升序数组
                // 所以当left的某一元素大于right数组的某一个元素，就说明left的该元素以及后面的元素都比right的该元素大
                count += left.length - i 
            }
        }

        while(i < left.length) {
            res.push(left[i])
            i++
        }

        while(j < right.length){
            res.push(right[j])
            j++
        }
    
        return res
    }
    return count
};


console.log(`数组${arr}中,逆序对的总数:${reversePairs(arr)}`)

File: Merge-Sort/python/leetcode_lc_51.py
nums = [7,5,6,4]

class Solution:
    count=0
    def reversePairs(self, nums):
        self.mergeSort(nums)
        return self.count

    def mergeSort(self,nums):
        if len(nums) < 2:
            return nums
        
        mid = int(len(nums)/2)
        left = nums[0:mid]
        right = nums[mid:]

        return self.merge(self.mergeSort(left),self.mergeSort(right))

    def merge(self,left,right):
        res = []
        i = j = 0
        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                res.append(left[i])
                i += 1
            else:
                for idx in range(i,len(left)):
                    print("***逆序对:",left[idx],right[j])
                res.append(right[j])
                j += 1
                self.count += len(left) - i

        while i < len(left):
            res.append(left[i])
            i += 1

        while j < len(right):
            res.append(right[j])
            j += 1

        return res


solution = Solution()

print("数组{}中,逆序对的总数:{}".format(nums,solution.reversePairs(nums)))